### ISSUE TITLE
TSTyche setup issue

### ISSUE THREAD

**Issue by mrazauskas** [ghallyb/ts-binary-array#1]
First of all, glad to see you are using TSTyche! This is my project.

Just wanted to draw your attention that there is serious setup issue. The TSConfig file you specified does not include the `./tmp` directory and excludes all test files as well:

https://github.com/ghallyb/ts-binary-array/blob/057e84454ca6068639f5a2c1fa04992323173ba6/tsconfig.json#L12-L13

This does not work. TSTyche loads options from `tsconfig.json` only if it includes a file in question. See documentation: https://tstyche.org/reference/config-file#tsconfig

**Comment by ghallyb** [comment]
Thanks for raising the issue. I appreciate you taking a look. Interestingly, the setup is working correctly for what I am trying to do, but on the face of it, it looks like it shouldn't be (based on what you have said). 

I will have a play around and try and work out what is going on.

**Comment by ghallyb** [comment]
Ah yes, I see what you mean from the docs. The tsconfig isn't being used at all and default options are being used instead.

I will be adding a correctly configured `tsconfig.test.json` and using that within my `tstyche.config.json`. I will keep the other `tsconfig.json` specifically for building the project (but not testing).

**Comment by mrazauskas** [comment]
Yep, that is exactly what I wanted to suggest. Would be even better to keep test files in a separate directory, but that is up to you (;

**Comment by mrazauskas** [comment]
Alternatively you could have `./tsconfig.json` that includes all the file and a dedicated `./tsconfig.build.json`.

In this case your IDE and TSTyche would use the same TSConfig. This also means that all `*.test.ts` files would also be included in the `./tsconfig.json` project. Currently your IDE falls back to its defaults.

**Comment by ghallyb** [comment]
I appreciate the advice. I will have a think on the best way to set it up (still messing around with things just now).

With regards to having a separate directory (or directories) for tests:
- Usually I subscribe to the idea that tests should live next to their respective implementations.
- However, this is the first time that I'm doing type testing as well as functional testing and I find that having both test.ts and tst.ts files next to the implementations is somewhat messy. To that end, I probably will move the tests into their own directory at some point.

I am interested in a couple of things in relation to your project:

**How did you know my project was using yours?**
Is that a github feature, or npm or something else? Just curious really.

**Affirmative Error Checking**
I need to test that some scenarios  _**do**_ raise an error. As such I am using the `x.type.toRaiseError()` to check this. The problem with this is that then the test files show up as red in the IDE because the test scenarios obviously have deliberate errors in them (see screenshot).

![type-error](https://github.com/user-attachments/assets/911ecaaf-7999-4795-a668-6c0a4a620f44)

As a workaround, I am doing the following:
- Use `@ts-ignore` comment to make sure these errors don't show up in the ide
- When running type tests, I use a script that does the following.
- Copy the whole file structure to `tmp`
- Find and replace the `@ts-ignore` (so that the deliberate errors are now visibile to tsc)
- Run tstyche on `tmp`
- Delete `tmp` afterwards

I notice some people raised this (or a similar issue) to you and you recommended that the following feature in [this PR](https://github.com/tstyche/tstyche/pull/52 ) would probably solve the problem. I think you are correct in that regard, so I was just wondering:
- Is this feature still in development or have you run into a roadblock with it?
- Is there a github issue or document that fleshes out how the api is going to work?
- If it is still in development, would you be interested in some assistance with it (assuming I can be of any help)?

**Comment by mrazauskas** [comment]
Thanks! This is very interesting to talk about.

By the way, did you consider keeping functional and type tests in a single file? That could be another solution to reduce the number of files: https://tstyche.org/guide/test-runner#functional-tests

Yes, GitHub lists dependents of each library: https://github.com/tstyche/tstyche/network/dependents

I keep an eye on that list to see how TSTyche is used. What could be improved or what is missing in the library? Like you noticed `.toBeCallableWith()` is obviously needed.

It will work the same as calling a function with given arguments. `.toAcceptProps()` is implemented similarly. The difference is that call-ability is very broad topic. Overloads and optional args are the easy parts. But rest params and especially generic functions are complex to deal with. To implement everything at once simply requires a lot of time. So I think to work on the easy parts first and to add other features gradually.

**Comment by ghallyb** [comment]
Ah nice. I hadn't noticed the the dependents list. 

Also, I did not know I could mix functional and type tests together. I may evaluate that approach.

With regards to `toBeCallableWith()`, having this would solve the issue I have with the screenshot above. 

Instead of writing this (which errors out in the IDE without the `@ts-ignore`):
```typescript
    it("the second argument should reject any value that isn't one of 8 | 16 | 32", () => {
        const len = 12; 
        // @ts-ignore
        expect(new FixedBinaryArray(len, 0)).type.toRaiseError();
   })
```
I could write it as follows (which if I understand correctly, wouldn't cause an error in the IDE):
```typescript
    it("the second argument should reject any value that isn't one of 8 | 16 | 32", () => {
        const len = 12; 
        expect(FixedBinaryArray).type.not.toBeCallableWith(len, 0);
   })
```

With regards to the other items from the PR that aren't done yet:
![image](https://github.com/user-attachments/assets/0b1ec67e-94be-4302-b6c7-91a1ba9b3bb3)

I was mostly just curious as to what their use cases are. If I think of anything specific as I write more tests, I can feed it back to you.

**Comment by mrazauskas** [comment]
Right, code editor will not show any errors. That’s the goal!

Regarding the use cases, it is very easy to mixup oder of overloads, so testing could ensure all works as expected.

Also here is a type that has rest params (would be good to test that it requires at least one `key`): https://github.com/tstyche/tstyche/blob/831e38413d64c2fd4f6fea189fe96348bfc981cf/source/types.ts#L319

And so on (;

**Comment by mrazauskas** [comment]
Quick update: the `.toBeConstructableWith()` matcher got added to TSTyche recently and it ships with latest beta. Simply install: `tstyche@next`.

More details in the release: https://tstyche.org/releases/tstyche-4#to-be-constructable-with
